### DISCUSSION TITLE
AP_IF DHCP leaking onto STA_IF? port-rp2

### DISCUSSION THREAD

**Discussion by simonmccaughey** [micropython/micropython#9599] [Core Development]
This may be a red herring, but just raising it in case someone else sees the same. **_Edit to add, not a red herring, it's an issue. DHCP is leaking across the interfaces, which should be independent_** 

The thermostats in my house are custom code which was previously running on ESP8266, and recently converted to Pi Pico, and suddenly I get some devices connected to the LAN from my router with 192.168.4.16 (or similar addresses) - which then obviously cannot talk to the internet using my normal router.

This has happened to both a LAN connected PC and a hardware device for my solar, both of which I could see getting the '4' address range. Seems like an odd coincidence, but also seems unlikely. When I unplug my thermostats and reboot the devices they come up with addresses in the normal range

**Comment by robert-hh** [comment]
Did you disable STA_AP mode of your device? Maybe it's still on after having it enabled earlier.

**Reply by simonmccaughey** [reply]
> Did you disable STA_AP mode of your device? Maybe it's still on after having it enabled earlier.

Sorry, I should have been clear, I want the AP active, so I have the option to connect that way to configure it - but the 192.168.4.* address range shouldn't 'leak' onto the STA_IF interface

**Reply by robert-hh** [reply]
You can connect as well to the device when it is in STA mode. No need to run it in AP mode.

**Reply by simonmccaughey** [reply]
> You can connect as well to the device when it is in STA mode. No need to run it in AP mode.

I run AP mode for initial setup, to connect the device to the WiFi using a web interface, plus its easy to connect if you don't know the IP address. Either way, the DHCP allocations from AP mode shouldn't appear on the other network interface, right? I should be able to have AP mode on without it interfering with the network connected to the other interface. Both interfaces should be independent.

**Reply by robert-hh** [reply]
You can give your board a fixed IP address for station mode. Then you know the address. If you still want to run the board in AP mode, you have to use a different IP address range. But that does not seem to be appropriate for you. And I see no option to stop the DHCP service.
As you have noticed, it gets a problem if two DHCP services are running at the same network address range. Both will service requests for an address, and the first one responding wins.

**Reply by simonmccaughey** [reply]
> You can give your board a fixed IP address for station mode. Then you know the address. If you still want to run the board in AP mode, you have to use a different IP address range. But that does not seem to be appropriate for you. And I see no option to stop the DHCP service. As you have noticed, it gets a problem if two DHCP services are running at the same network address range. Both will service requests for an address, and the first one responding wins.

I think we have our wires crossed here, there are not two DHCP services running in the same network range. STA_IF is on my home WiFi, running the 192.168.2.* range, and AP_IF is running 192.168.4.* range. They should be isolated, as they are two separate network interfaces on the device. But other devices connected to my 192.168.2.1 router, via LAN ports are receiving DHCP addresses in the 192.168.4.* range. That should not be happening.

**Reply by simonmccaughey** [reply]
To make it clearer, I made a diagram

![image](https://user-images.githubusercontent.com/39386195/195570779-58353b36-9dd4-4eb4-bd91-93200f7efc3c.png)

On the connected linux box, DHCP requests sometimes get answered by the pico, and sometimes by the router. This is new behaviour since I switched from ESP8266

See below, the same DHCP command **_ran on the linux box_** gets two different responses.

```
pi@pix:~ $ sudo nmap --script broadcast-dhcp-discover -e eth0
Starting Nmap 7.70 ( https://nmap.org ) at 2022-10-13 11:08 BST
Pre-scan script results:
| broadcast-dhcp-discover:
|   Response 1 of 1:
|     IP Offered: 192.168.2.199
|     DHCP Message Type: DHCPOFFER
|     Server Identifier: 192.168.2.1
|     IP Address Lease Time: 2m00s
|     Renewal Time Value: 1m00s
|     Rebinding Time Value: 1m45s
|     Subnet Mask: 255.255.255.0
|     Broadcast Address: 192.168.2.255
|     Router: 192.168.2.1
|     Domain Name Server: 192.168.2.1
|     Domain Name: broadband
|_    NTP Servers: 192.168.2.1
WARNING: No targets were specified, so 0 hosts scanned.
Nmap done: 0 IP addresses (0 hosts up) scanned in 3.18 seconds
pi@pix:~ $ sudo nmap --script broadcast-dhcp-discover -e eth0
Starting Nmap 7.70 ( https://nmap.org ) at 2022-10-13 11:14 BST
Pre-scan script results:
| broadcast-dhcp-discover:
|   Response 1 of 1:
|     IP Offered: 192.168.4.16
|     DHCP Message Type: DHCPOFFER
|     Server Identifier: 192.168.4.1
|     Subnet Mask: 255.255.255.0
|     Router: 192.168.4.1
|     Domain Name Server: 192.168.4.1
|_    IP Address Lease Time: 1d00h00m00s
WARNING: No targets were specified, so 0 hosts scanned.
Nmap done: 0 IP addresses (0 hosts up) scanned in 3.14 seconds
```

**Reply by robert-hh** [reply]
Since the DHCP discover message is a broadcast, all DHCP servers will respond, and the client will accept the first DHCP offer it receives.

**Reply by simonmccaughey** [reply]
> Since the DHCP discover message is a broadcast, all DHCP servers will respond, and the client will accept the first DHCP offer it receives.

Yes, agreed! But there should be no DHCP server on the pico STA_IF, 192.168.2.162 in my diagram. STA_IF should only be a client on the network, and not answer a DHCP broadcast. There should only be one DHCP server on the router network, the router one. To access the DHCP server on the pico, you should need to connect via the AP.

Why is the pico answering DHCP requests on its client STA_IF interface? Surely that is a bug

**Reply by robert-hh** [reply]
In your log I see two DHCP responses. One from the router and one from the Pico. Both respond because they receive the DHCP discover message, which is not for a specific network..
